datafusion/src/physical_plan/hash_join.rs

```diff
@@ -476,18 +483,18 @@ fn update_hash(
 
     // insert hashes to key of the hashmap
     for (row, hash_value) in hash_values.iter().enumerate() {
-        match hash.raw_entry_mut().from_hash(*hash_value, |_| true) {
-            hashbrown::hash_map::RawEntryMut::Occupied(mut entry) => {
-                entry.get_mut().push((row + offset) as u64);
-            }
-            hashbrown::hash_map::RawEntryMut::Vacant(entry) => {
-                entry.insert_hashed_nocheck(
-                    *hash_value,
-                    (),
-                    smallvec![(row + offset) as u64],
-                );
-            }
-        };
+        let item = hash_map
+            .0
+            .get_mut(*hash_value, |(hash, _)| *hash_value == *hash);
```

Comment:
I don't fully understand how this works in the case of hash collisions. Don't you also have to check if the existing values in the table (aka the `_`) are the same as new row that came in to ensure two rows don't have the same hash values?

Although now that I write this, I am not sure how the existing structure handles hash collisions either as it seems to only look at the hash values and not the values in the rows themselves. 

I found creating a test for collisions  incredibly hard, btw, as I couldn't actually find any `u32` values that collided and I blew out the memory on my machine searching for `u64` values that collided 😆

Reply:
> I don't fully understand how this works in the case of hash collisions. Don't you also have to check if the existing values in the table (aka the _) are the same as new row that came in to ensure two rows don't have the same hash values?

This doesn't check for actual values but AFAIK is required to avoid mapping/finding different hash values in the same bucket and finding the wrong bucket for a different hash. I tested it earlier and it seems the case should also be extremely rare too (couldn't get a single collission within 100M values). Problem of having the hash values in a same bucket is that another `u64` hash could point to the wrong bucket in the probing phase. 
I couldn't find a way yet to test this properly. The case of hash collissions by having the offsets present in different `u64` has a unit test. 

See here for the note where it says that there could be a same bucket for different hashes:
https://docs.rs/hashbrown/0.11.2/src/hashbrown/raw/mod.rs.html#1043-1053 

Also the `u64` output of the hasher seem of high quality so very hard to get collissions there indeed (at least for single columns, with multiple columns `combine_hashes` also affects the quality).
The actual case for detecting whether there is a colission of the values for the offsets happens at the moment elsewhere in the code.

I am not sure whether it makes sense to do the "value equality" check inside the `eq` parameter of the insert for the hash join. Probably it will only add time if there are not a lot of actual `u64` collissions, as it should be done in the probing too anyway.